### ISSUE TITLE
@wordpress/stan Follow-up tasks

### ISSUE THREAD

**Issue by youknowriad** [WordPress/gutenberg#27109]
In #26866 we're introducing a new package called `@wordpress/stan` with its bindings with `@wordpress/data`. Read the PR description and the package README for more details.

Introducing a new way to register stores and deal with state is a big task and this issue is going to serve as a follow-up todo list:

 - [x] Try use-subscription for use-select (check the performance results) https://github.com/WordPress/gutenberg/pull/26866#discussion_r523978318
 - [ ] Check how `@wordpress/data` bindings perform in concurrent mode https://github.com/WordPress/gutenberg/pull/26866#issuecomment-727926428
 - [ ] Introduce simple react bindings for `@wordpress/stan` for folks that don't want to use `@wordpress/data` (hook creator like zustand or provider based like jotai)
 - [ ] Build ReduxDevTools integration https://github.com/WordPress/gutenberg/pull/26866#issuecomment-728022810
 - [ ] Stabillize and document the `@wordpress/data` bindings.
 - [ ] Decide whether to keep stan as a bundled package or not.
 - [x] Replace the atom family concept with a more generic selector concept. https://github.com/WordPress/gutenberg/pull/27147

--- 
In parallel, we can also start trying to refactor existing stores to atomic stores. This work is expected to have good impact on performance especially for central stores like `block-editor`

 - [ ] core (entities)
 - [ ] core/data (resolution state)
 - [ ] core/block-editor
 - [ ] core/editor
 - [ ] core/viewport
 - [ ] core/edit-post
 - [ ] core/edit-widgets
 - [ ] core/edit-site
 - [ ] core/blocks @youknowriad #27140

Feel free to own one of these tasks.

**Comment by youknowriad** [comment]
I've tried use-subscription unsuccessfully, it rerenders too often (no shallow equal comparison) and breaks several of our tests. Maybe I'm not integrating it properly but I consider this to-do list item done.

**Comment by adamziel** [comment]
Surfacing this discussion about error handling https://github.com/WordPress/gutenberg/pull/26866#discussion_r527737511 and my async/await explorations here: https://github.com/WordPress/gutenberg/pull/27155

**Comment by adamziel** [comment]
I spent some more time thinking about this and I stumbled upon the following problems:

### How should "concurrency" work?

Imagine the following scenario:

```js
	const sleep = ( ms ) => new Promise( ( resolve ) => setTimeout( resolve, ms ) );
	const count1 = createAtom( 1 );
	const count2 = createAtom( 1 );
	const asyncCount = createDerivedAtom( async ( { get } ) => {
		const _count1 = get( count1 );
		await sleep( 1000 );
		const _count2 = get( count2 );
		return _count1 + _count2;
	} );
	const registry = createAtomRegistry();

	registry.subscribe( asyncCount, () => {} );
	await jest.advanceTimersByTime( 1000 );
	await flushImmediatesAndTicks();

	expect( registry.get( asyncCount ) ).toEqual( 2 );
	registry.set( count1, 5 );
	await jest.advanceTimersByTime( 500 );
	await flushImmediatesAndTicks();

	registry.set( count1, 9 );
	await jest.advanceTimersByTime( 500 );
	await flushImmediatesAndTicks();
	expect( registry.get( asyncCount ) ).toEqual( /* ? */ );
```

What will `asyncCount` be? Well, the resolver will be triggered three times: 1. initially, 2. when `count1` is set to 5, 3. when `count1` is set to 9. Thus, `asyncCount` will be changed over time with the following values: `2` -> `6` -> `10`. By itself, this makes sense as a model. But it creates a number of problems as I'll show below.

### How should dependencies of asynchronous resolvers be tracked?

Let's make the following change to the code snippet above:

```js
	const asyncCount = createDerivedAtom( async ( { get } ) => {
		const _countA = get( count1 );
		let _countB;
		if ( _countA === 5 ) {
			await sleep( 1000 );
			_countB = get( count3 );
		} else {
			await sleep( 200 );
			_countB = get( count2 );
		}
		return _countA + _countB;
	} );
```

#### What is the value of `asyncCount`?

It's `6` in the line `expect( registry.get( asyncCount ) ).toEqual( /* ? */ );`.

I checked what Recoil does in an analogous situation, and it discards the value that started resolving earlier. So basically in Recoil we'd see `asyncCount` to have value `10` which I believe is the correct outcome.

#### What are the dependencies of `asyncCount`?

In the current imlementation, they are `[ count1, count3 ]`.

I checked what happens in Recoil. It seems like it uses immutable dependency trees and simply throws an error in the following case:

```js
selector( {
	key: 'myAsyncState',
	get: async ( { get } ) => {
		get( otherState1 );
		await sleep(1000);
		get( otherState2 ); // Error: Tried to save dependencies to a discarded tree
	},
} );
```

### How should async errors behave?

Let's now make our resolver throw an error:

```js
	const asyncCount = createDerivedAtom( async ( { get } ) => {
		const _countA = get( count1 );
		let _countB;
		if ( _countA === 5 ) {
			await sleep( 1000 );
			_countB = get( count3 );
			throw new Error();
		} else {
			await sleep( 200 );
			_countB = get( count2 );
		}
		return _countA + _countB;
	} );
```

Notice that the error is only thrown once a newer `asyncCount` is already resolved. What should happen here? I'm not sure. I noticed Recoil just swallows the error as if nothing happened. In contrast, if there is no newer version available, the error is made available to the React component.

### Possible solutions

These are just some of the issues I identified on the spot, I'm sure there's much more waiting for us to uncover. 

One way forward would be to keep discussing and addressing these problems one by one in `@wordpress/stan`. My fear here is that we'd be just rewriting Recoil since it's meant to solve exactly the same problems. It is unfortunate that it's tightly coupled to React and thus not a good fit here. I wonder if there's any way to reuse some of its parts?

Another solution would be changing `@wordpress/stan` API a bit so that it's capable of playing nicely with with RFP library like `Rx.js`. From there, it should be quite easy to both reason about the data flow and also augment it as we need.

**Comment by jsnajdr** [comment]
I don't have any definitive answers, just an idea how to formulate the problem.

When executing this code snippet:
```js
const _count1 = get( count1 );
await sleep( 1000 );
const _count2 = get( count2 );
```
The figuring out what it means all boils down to whether we're working with a constant state snapshot the entire time, or if we're retrieving values from the "current state" at any time.

Imagine we're working with something like Redux state. Then we can implement `get` in two ways:
```js
const state = store.getState();
const get = atom => state[ atom ];
resolver( { get } );
```
or
```js
const get = atom => getState()[ atom ];
resolver( { get } );
```

The first option looks more tractable: the entire derived atom tree is calculated from one state snapshot, not from multiple ones. If we used multiple snapshots, then determining which snapshots exactly are used is basically random.

**Comment by adamziel** [comment]
One more factor to consider is the new React concurrent mode that someone (was it you @jsnajdr?) surfaced in another thread and the "tearing" problem it creates. It seems like its more or less the same issue I outlined above, only caused on a higher level:

https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-mode

**Comment by youknowriad** [comment]
Head's up that I'm going to  put this on hold for a bit until we measure with certainty the performance impact on block-editor.  Otherwise, it's just a big refactor for subjective devx impact.

In the meantime, I'm continuing my performance checks.

**Comment by adamziel** [comment]
@jsnajdr spot on formulation! I agree this is essentially what we're dealing with on per-selector level. I would generalize it even more and avoid inconsistencies not only on the atom/selector level, but also across different components in the entire React. I'll abstain from exploring this further from now but it could mean we'd have to somehow commit state updates in batches.

**Comment by grzim** [comment]
Rx.js solves problems with concurrency by design - working with observables only, makes the entire problem with async go away - we treat synchronous and asynchronous sources the same. The library is well maintained, documented, and tested so choosing it could be more futureproof and less expensive when it comes to maintenance. Another thing is typings - rx comes with types, which is super helpful if we would like to gradually introduce TypeScript to the project. And last but not least - as it is a quite popular solution (especially among Angular folks) it will be easier for people knowing it to contribute, rather than learning a custom solution from scratch.
When it comes to disadvantage - it is hard to understand for people not accustomed to reactive thinking. In many Angular projects, where people are forced to use rx it is often a case it is hacked in billion ways to make it work. Here it would work similarly I guess, so people should be able to use some imperative API. Creating an imperative API on top of Rx would be a painful job though, as rx is made by design, to avoid such an approach. 
When it comes to resolvers API, it is possible to create observables from generators, but for sure making resolvers functional would be the best approach.

**Comment by youknowriad** [comment]
I feel right now, that there's too much work needed here for something that is not a priority. This means that I won't get to it anytime soon. Folks feel free to chime in and do your own explorations.
